logexp/cmd/main.go

```diff
@@ -0,0 +1,257 @@
+		return fmt.Errorf("Writing to file of journalctl logs for '%v' service failed: %v", service, err)
+	}
+	return nil
+}
+
+// Create logfiles for systemd services in outputDir, based on whether the node isMaster.
+func createSystemdLogfiles(outputDir string, isMaster bool) {
+	services := systemdServices
+	if ! isMaster {
+		services = append(services, nodeSystemdServices...)
+	}
+	for _, service := range services {
+		if err := createSystemdLogfile(service, "cat", outputDir); err != nil {
+			glog.Warningf("Failed to record journalctl logs: %v", err)
+		}
+	}
+	// Service logs specific to VM setup.
+	for _, service := range systemdSetupServices {
+		if err := createSystemdLogfile(service, "short-precise", outputDir); err != nil {
+			glog.Warningf("Failed to record journalctl logs: %v", err)
+		}
+	}
+}
+
+// Copy logfiles specific to this node based on its type, cloud-provider, system services, etc
+// to a temporary directory. Also create logfiles for systemd services if journalctl is present.
+// We do not expect this function to see an error.
+func prepareLogfiles(logDir string) {
+	glog.Infof("Preparing logfiles relevant to this node")
+	logfiles := make([]string, 0)
+	var isMaster bool
+
+	// Select node-type (master/node) specific logs.
+	if _, err := os.Stat(localLogPath + "/kube-apiserver.log"); err == nil {
+		glog.Infof("VM identified as kubernetes master")
+		isMaster = true
+		logfiles = append(logfiles, masterLogs...)
+	} else {
+		glog.Infof("VM identified as kubernetes node")
+		isMaster = false
+		logfiles = append(logfiles, nodeLogs...)
+	}
+
+	// Select cloud-provider specific logs.
+	switch kubernetesProvider {
+	case "gce", "gke":
+		logfiles = append(logfiles, gceLogs...)
+	case "kubemark":
+		logfiles = append(logfiles, gceLogs...)  // TODO(shyamjvs): Make this variable wrt kubemark's real provider.
+		if ! isMaster && enableHollowNodeLogs {
+			logfiles = append(logfiles, kubemarkLogs...)
+		}
+	case "aws":
+		logfiles = append(logfiles, awsLogs...)
+	default:
+		glog.Warningf("Unknown cloud provider '%v' provided, skipping any provider specific logs", kubernetesProvider)
+	}
+
+	// Select system/service specific logs.
+	if _, err := os.Stat("/workspace/etc/systemd/journald.conf"); err == nil {
+		glog.Infof("Journalctl found on host. Collecting systemd logs")
+		createSystemdLogfiles(logDir, isMaster)
+	} else {
+		glog.Infof("Journalctl not found on host (%v). Collecting supervisord logs instead", err)
+		logfiles = append(logfiles, kernelLog)
+		logfiles = append(logfiles, initdLogs...)
+		logfiles = append(logfiles, supervisordLogs...)
+	}
+
+	// Copy all the logfiles that exist, to logDir.
+	for _, logfile := range logfiles {
+		logfileFullPath := filepath.Join(localLogPath, logfile + ".log*")  // Append .log* to copy rotated logs too.
+		cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("cp %v %v", logfileFullPath, logDir))
+		if _, err := cmd.Output(); err != nil {
```

Comment:
Seems like every call to Wait should be preceded by a call to Start. However, [Run](https://golang.org/pkg/os/exec/#Cmd.Run) seems to be the option we are looking for. Changed it at all relevant places.